correctly close streams on error conditions

dryajov opened the pull request:
This is some more cleanup as well as handling connection/stream cleanup properly in a few more places.

dryajov committed:
fix: close streams on stop error

dryajov committed:
feat: close stream on invalid message & cleanup

dryajov committed:
fix: proper stream closing

dryajov committed:
fix: hop tests to reflect latests changes

jacobheun commented on src/circuit/hop.js:
```diff
@@ -106,11 +107,12 @@ class Hop extends EE {
       if (err) {
         log.err(err)
-        setImmediate(() => this.emit('circuit:error', err))
+        sh.close()
```
I ran the libp2p, switch and interop tests for this update and switch was the only thing that found an issue.

This test https://github.com/libp2p/js-libp2p-switch/blob/v0.40.8/test/circuit-relay.node.js#L322-L329 had the issue.

What's happening is when the circuit dials to the peer, it gets the muxer error for not supporting relay. This causes the stream to be closed at https://github.com/libp2p/js-libp2p-circuit/blob/27593926fa221d8c520d9dfd227fb47d31e82f01/src/circuit/hop.js#L213. That error is passed to L112, where the stream is closed again. Since it was already closed, there is no longer a shake in the stream handler, https://github.com/libp2p/js-libp2p-circuit/blob/27593926fa221d8c520d9dfd227fb47d31e82f01/src/circuit/stream-handler.js#L111, so `.rest` throws an error:

```
     Uncaught TypeError: Cannot read property 'rest' of null
      at StreamHandler.rest (node_modules/libp2p-circuit/src/circuit/stream-handler.js:111:29)
      at StreamHandler.close (node_modules/libp2p-circuit/src/circuit/stream-handler.js:128:12)
```

We should add a test to this repo to replicate the issue so we catch future issues.

dryajov replied:
Thanks for catching this. I'll add a fix and the test case 👍

dryajov replied:
@jacobheun I've pushed the fixes for the `SH.close()` issues - it now allows to be called more than once without crashing. 

I've also added missing cb returns in stream handler, that should prevent it from going into a bad state. This should address the issues reported in:

- https://github.com/libp2p/js-libp2p-circuit/issues/36
- https://github.com/libp2p/js-libp2p-circuit/issues/35
- https://github.com/libp2p/js-libp2p-circuit/issues/34

However, I realize that whats needed are stream-handler tests, which are currently missing, I'd love to add them in, but I'm a bit low on time right now. I'll open up an issue to keep track of that.

dryajov committed:
fix: streamhandler close and return CBs on error

- fixed StreamHandler.close to allow calling more than once without
crashing
- return callback on error, to stop further executiion when SH is in an invalid state

dryajov commented:
Issue to track missing StreamHandler tests - https://github.com/libp2p/js-libp2p-circuit/issues/37

jacobheun commented:
Tests look good across libp2p, ipfs/interop and libp2p-switch! Thanks for opening #37.
